core/primitives/src/rand.rs

```diff
@@ -0,0 +1,187 @@
+    }
+
+    pub fn sample(&self, seed: [u8; 32]) -> usize {
+        let usize_seed = Self::copy_8_bytes(&seed[0..8]);
+        let balance_seed = Self::copy_16_bytes(&seed[8..24]);
+        let uniform_index = usize::from_le_bytes(usize_seed) % self.aliases.len();
+        let uniform_weight = Balance::from_le_bytes(balance_seed) % self.weight_sum;
+
+        if uniform_weight < self.no_alias_odds[uniform_index] {
+            uniform_index
+        } else {
+            self.aliases[uniform_index] as usize
+        }
+    }
+
+    pub fn get_aliases(&self) -> &[u64] {
+        &self.aliases
+    }
+
+    pub fn get_no_alias_odds(&self) -> &[Balance] {
+        &self.no_alias_odds
+    }
+
+    fn copy_8_bytes(arr: &[u8]) -> [u8; 8] {
+        let mut result = [0u8; 8];
+        result.clone_from_slice(arr);
+        result
+    }
+
+    fn copy_16_bytes(arr: &[u8]) -> [u8; 16] {
+        let mut result = [0u8; 16];
+        result.clone_from_slice(arr);
+        result
+    }
+}
+
+/// Sub-module to encapsulate helper struct for managing aliases
+mod aliases {
+    pub struct Aliases {
+        aliases: Vec<usize>,
+        smalls: Vec<usize>,
+        bigs: Vec<usize>,
+    }
+
+    impl Aliases {
+        pub fn new(n: usize) -> Self {
+            let mut aliases = Vec::with_capacity(n);
+            for _ in 0..n {
+                aliases.push(0);
+            }
+            Self { aliases, smalls: Vec::with_capacity(n), bigs: Vec::with_capacity(n) }
+        }
+
+        pub fn push_big(&mut self, b: usize) {
+            self.bigs.push(b);
+        }
+
+        pub fn pop_big(&mut self) -> usize {
+            self.bigs.pop().unwrap()
+        }
+
+        pub fn bigs_is_empty(&self) -> bool {
+            self.bigs.is_empty()
+        }
+
+        pub fn push_small(&mut self, s: usize) {
+            self.smalls.push(s);
+        }
+
+        pub fn pop_small(&mut self) -> usize {
+            self.smalls.pop().unwrap()
+        }
+
+        pub fn smalls_is_empty(&self) -> bool {
+            self.smalls.is_empty()
+        }
+
+        pub fn set_alias(&mut self, index: usize, alias: usize) {
+            self.aliases[index] = alias;
+        }
+
+        pub fn get_aliases(self) -> Vec<u64> {
+            self.aliases.into_iter().map(|a| a as u64).collect()
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::hash;
+    use crate::rand::WeightedIndex;
+
+    #[test]
+    fn test_should_correctly_compute_odds_and_aliases() {
+        // Example taken from https://www.keithschwarz.com/darts-dice-coins/
+        let weights = vec![5, 8, 4, 10, 4, 4, 5];
+        let weighted_index = WeightedIndex::new(weights);
+
+        assert_eq!(weighted_index.get_aliases(), &[1, 0, 3, 1, 3, 3, 3]);
+
+        assert_eq!(weighted_index.get_no_alias_odds(), &[35, 40, 28, 29, 28, 28, 35]);
+    }
+
+    #[test]
+    fn test_sample_should_produce_correct_distribution() {
+        let weights = vec![5, 1, 1];
```

Comment:
This test fails for [4, 6, 1, 1] (with asserts properly updated).

If you look at the aliases section in the page you linked (https://www.keithschwarz.com/darts-dice-coins/), this is exactly the example on which he explains it, but with 6 and 4 swapped. If I don't swap them (leave as 6, 4, 1, 1), your code works correctly (it 6,4,1,1 -> 4,4,3,1 -> 3,4,3,2 -> 3,3,3,3), but if I do, you hit the issue he explains at the very beginning of the section (4,6,1,1 -> 3,6,1,2 -> 3,4,3,2 -> ???).

`big` needs to be a maxheap.

Also, the test itself is extremely flacky (fails with relative error 0.7%-0.8% one out of 5 executions). Though I ran it with `rand::thread_rng` instead of hashes, may be it changes the distribution.